docs/HTTP_V2_SPEC.md

```diff
@@ -1,106 +1,236 @@
-executed as if a fresh Hrana stream was opened for the request.
+The HTTP API uses data structures and semantics from the Hrana 2 protocol.
 
-All request and response bodies are encoded in JSON, with content type
-`application/json`.
+Individual requests on the same stream are tied together by the use of a baton.
+The server returns a baton in every response to a request on the stream, and the
+client then needs to include the baton in the subsequent request. The client
+must serialize the requests: it must wait for a response to the previous request
+before sending next request.
 
-## Version check
+The server can also optionally specify a different URL that the client should
+use for the requests on the stream. This can be used to ensure that stream
+requests are "sticky" and reach the same server.
 
-```
+The server will close streams after a short period of inactivity, to make sure
+that abandoned streams don't accumulate on the server.
+
+## Check support for version 2
+
+```typescript
 GET /v2
 ```
 
-A server that supports HTTP API v2 must return a success response for requests
-on path `/v2`. This can be used as a crude and unreliable mechanism for clients
-to fall back to version 1 of the protocol if the server does not support version
-2.
+If the server supports this version of the HTTP API, it should return a 2xx
+response for a GET request on `/v2`. This can be used as a crude version
+negotiation mechanism by the client.
 
-## Execute a statement
+## Execute requests on a stream
 
-```
-POST /v2/execute
+```typescript
+POST /v2/stream
 
 -> {
-    "stmt": Stmt,
+    "baton": string | null,
+    "requests": Array<StreamRequest>,
 }
 
 <- {
-    "result": StmtResult,
+    "baton": string | null,
+    "base_url": string | null,
+    "results": Array<StreamResult>
+}
+
+type StreamResult =
+    | StreamResultOk
+    | StreamResultError
+
+type StreamResultOk = {
+    "type": "ok",
+    "response": StreamResponse,
+}
+
+type StreamResultError = {
+    "type": "error",
+    "error": Error,
 }
 ```
 
-The `execute` endpoint receives a statement and returns the result of executing
-the statement. The `Stmt` and `StmtResult` structures are from the Hrana
-protocol. The semantics of this endpoint is the same as the `execute` request in
-Hrana.
+The `stream` endpoint is used to execute requests on a stream. `baton` in the
+request specifies the stream. If the client sets `baton` to `null`, the server
+should create a new stream.
+
+Server responds with another `baton` value in the response. If the `baton` value
+in the response is `null`, it means that the server has closed the stream. The
+client must use this value to refer to this stream in the next request (the
+`baton` in the response should be different from the `baton` in the request).
+This forces the client to issue the requests serially: it must wait for the
+response from a previous `stream` request before issuing another request on the
+same stream.
```

Comment:
as per my above example, I don't see why we need to bother with baton if the client side can trust serialization on its side, either by waiting on previous requests or natively if the language doesn't support any parallel execution.

just makes the client implementation harder.

Reply:
The baton is used to identify the stream, while the `base_url` is just an optional mechanism to ensure that sessions are sticky. As Athos pointed out, we may use a different approach in the future, so it makes sense to implement these orthogonal features with orthogonal mechanisms.